### PR TITLE
Feat: Upgrade  django-oscar to version 3.2

### PR DESCRIPTION
**Pull request by zubair-ce07**
## Description
Changes related to upgrade ecommerce application to django-oscar  version 3.2 from 2.2

## Required Testing
Thoroughly test ecommerce dashboard. (Every flow needs to be tested)

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Other information
Ticket: https://2u-internal.atlassian.net/browse/REV-3639

### CODE REVIEW

File: ecommerce/templates/oscar/dashboard/catalogue/product_update.html [hunk 1]
```diff
@@ -223,8 +223,8 @@ <h3>{% trans "Upload, change or remove images" as tmsg %}{{ tmsg | force_escape
                                                     <td>{% include "oscar/dashboard/partials/form_field.html" with field=stockrecord_form.low_stock_threshold nolabel=True %}</td>
                                                 {% endif %}
                                                 <td>{% include "oscar/dashboard/partials/form_field.html" with field=stockrecord_form.price_currency nolabel=True %}</td>
-                                                <td>{% include "oscar/dashboard/partials/form_field.html" with field=stockrecord_form.cost_price nolabel=True %}</td>
-                                                <td>{% include "oscar/dashboard/partials/form_field.html" with field=stockrecord_form.price_excl_tax nolabel=True %}</td>
+                                                <!-- <td>{% include "oscar/dashboard/partials/form_field.html" with field=stockrecord_form.cost_price nolabel=True %}</td> -->
```
Comment (julianajlk) [inline ecommerce/templates/oscar/dashboard/catalogue/product_update.html]: Curious, why is this portion commented out?

File: ecommerce/templates/oscar/dashboard/orders/order_list.html [hunk 2]
```diff
@@ -139,7 +138,6 @@ <h3><i class="icon-warning"></i> {% trans "Change order status" as tmsg %}{{ tms
         {% endblock %}
 
 
-        {% include "oscar/dashboard/orders/partials/bulk_edit_form.html" with status=active_status %}
```
Comment (julianajlk) [inline ecommerce/templates/oscar/dashboard/orders/order_list.html:142]: This bulk edit form is no longer needed here? Is this part of the upgrade or something else?

**Review by julianajlk — COMMENTED** [review]
Thanks for doing this work! Question - are the template changes included in this PR part of the upgrade? Were you able to see the changes locally/do they make sense?

**Comment by christopappas** [comment]
i see that the tests are failing with somenthing related to `option.OPTIONAL`

not sure if you've been digging into this yet, but, in my local ecommerce i see OPTIONAL as an attribute of the `AbstactOption` class

```
root@ecommerce:/edx/app/ecommerce/ecommerce# grep 'AbstractOption' -A 35  /edx/app/ecommerce/venvs/ecommerce/lib/python3.8/site-packages/oscar/apps/catalogue/abstract_models.py 
class AbstractOption(models.Model):
    """
    An option that can be selected for a particular item when the product
    is added to the basket.

    For example,  a list ID for an SMS message send, or a personalised message
    to print on a T-shirt.

    This is not the same as an 'attribute' as options do not have a fixed value
    for a particular item.  Instead, option need to be specified by a customer
    when they add the item to their basket.
    """
    name = models.CharField(_("Name"), max_length=128)
    code = AutoSlugField(_("Code"), max_length=128, unique=True,
                         populate_from='name')

    REQUIRED, OPTIONAL = ('Required', 'Optional')
```

looks like optional was deleted in this commit: https://github.com/django-oscar/django-oscar/commit/dd5a151af2a06bcd8e17a402a0a6f5ca3b14d17c#diff-6101f645bcf6dd2482680b01d02acf8a6010004e7a7730a9513e9cdc41165aff

i think for backwards compatibility, we may want to add OPTIONAL with text of `'Optional'` to https://github.com/openedx/ecommerce/blob/master/ecommerce/extensions/catalogue/models.py#L179

File: ecommerce/templates/oscar/dashboard/orders/order_list.html (hunk 2 above)
Comment (zubair-ce07) [reply ecommerce/templates/oscar/dashboard/orders/order_list.html:142]: Yes this has been removed in latest version

File: ecommerce/templates/oscar/dashboard/catalogue/product_update.html (hunk 1 above)
Comment (zubair-ce07) [reply ecommerce/templates/oscar/dashboard/catalogue/product_update.html]: I Will remove this commented code. cost_price field is removed in latest django-oscar version

**Comment by zubair-ce07** [comment]
> Thanks for doing this work! Question - are the template changes included in this PR part of the upgrade? Were you able to see the changes locally/do they make sense?

Yes Otherwise whole pages alignment were out of order. We need to upgrade template pages as per new upgraded version.

File: ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py [hunk 3]
```diff
@@ -32,5 +32,35 @@ class Migration(migrations.Migration):
     ]
 
     operations = [
+        migrations.AddField(
```
Comment (christopappas) [inline ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py]: was this the old migration we're adding to? why do we need to do this? this seems odd to me... i'm worried that changing a migration that has already been applied 1) would not have any effect 2) break the application of migrations

File: ecommerce/extensions/voucher/utils.py [hunk 4]
```diff
@@ -537,8 +537,9 @@ def create_new_voucher(code, end_datetime, name, start_datetime, voucher_type):
     if not isinstance(end_datetime, datetime.datetime):
         end_datetime = dateutil.parser.parse(end_datetime)
 
+    name = name[:128 - len(voucher_code)] + voucher_code
```
Comment (christopappas) [inline ecommerce/extensions/voucher/utils.py:540]: nice

**Review by christopappas — COMMENTED** [review]
Thanks for all this work Zubair. Before I 👍 , I'd like to understand the update to the old migration file and why that is necessary

File: ecommerce/extensions/basket/migrations/0016_alter_lineattribute_value.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 3.2.20 on 2023-08-24 10:28
```
Comment (christopappas) [inline ecommerce/extensions/basket/migrations/0016_alter_lineattribute_value.py:1]: so it's definitely weird that this migration and other other one are not gettign covered by codecov. these were also made a few months back. i wonder if there's somtehing about changes we have made along the way  that has prevented these migrations from being relevant? i honestly am not sure. if we delete this migrations from you branch, re run makemigrations, and these appear again , and then you push them up this branch and we still see this coverage issue, then i think it might be fine. i just want to make sure there isn't anything extra that was created along the way that is no longer relevant 🤷 -- honestly though, i'm not certain why this is the case

File: ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py [hunk 6]
```diff
@@ -4,21 +4,20 @@
 from django.db import migrations, models
 from oscar.core.loading import get_model
 
-Option = get_model('catalogue', 'Option')
```
Comment (christopappas) [inline ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py:7]: was this change still necessary based on the latest approach you described?

File: ecommerce/extensions/basket/migrations/0016_alter_lineattribute_value.py (hunk 5 above)
Comment (christopappas) [reply ecommerce/extensions/basket/migrations/0016_alter_lineattribute_value.py:1]: this is the case for so many of the migrations 🤔 -- that's so odd...

**Comment by christopappas** [comment]
you may need to squash your commits btw. looks like there's a commitlint error

**Review by christopappas — APPROVED** [review]
i'm giving you a 👍 because i think you have done enough due diligence and QA here so far. if none of the most recent comments I make you feel like something to change, and if you are comfortable with the stage of this, i would like to see us merge this and test extensively on stage 🚀

File: ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py (hunk 6 above)
Comment (zubair-ce07) [reply ecommerce/extensions/catalogue/migrations/0027_catalogue_entitlement_option.py:7]: Hy @christopappas. Yes this is necessary otherwise it will load the latest Option model (One created after running all the migrations) giving error "required field is not defined" because at this point required column is not created in our db.
